Add noop register hook

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60975
* #60663
* #60565
* **#60685**

Summary: This PR is part of a stack aimed to address #58512. 
Expose a noop `register_hook` method on `SavedTensor`. The purpose of this PR is to test error messages when the API is used incorrectly.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466224](https://our.internmc.facebook.com/intern/diff/D29466224)

Varal7 committed:
Add noop register hook

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Add noop register hook"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Add noop register hook"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -153,6 +153,21 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
+        "retain_graph=True if you need to backward through the graph a second time or "
+        "if you need to access saved variables after calling backward.");
+    } else {
+    TORCH_CHECK(false, "Calling register_hook on a tensor with value None is forbidden");
```
nit: tensor -> saved tensor?
since tensors can't have value of None

soulitzer commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -37,6 +37,8 @@ class TORCH_API SavedVariable {
   /// circular reference.
   Variable unpack(std::shared_ptr<Node> saved_for = nullptr) const;
 
+  void register_hooks();
```
Could add a comment saying that this is temporarily no op

soulitzer reviewed (approved):
Looks good, some small comments

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -153,6 +153,21 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
+    TORCH_CHECK(false,
+        "Calling register_hook on a saved tensor after it has been freed. "
+        "Saved intermediate values "
+        "of the graph are freed when you call .backward() or autograd.grad(). Specify "
```
nit: weird new line

Varal7 committed:
Update on "Add noop register hook"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466224).

Varal7 committed:
Update on "Add noop register hook"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466224](https://our.internmc.facebook.com/intern/diff/D29466224)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466224).

Varal7 committed:
Update on "Add noop register hook"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466224](https://our.internmc.facebook.com/intern/diff/D29466224)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466224).
